### ISSUE TITLE
On error getResponseObject() is not transformed

### ISSUE THREAD

**Issue by tkvedaras** [GenesisGateway/genesis_php#9]
```php
/** @var \Genesis\API\Response $response */
$response = $this->gateway->response();
$response->getResponseObject();
```

On Validation error `$response->getResponseObject();` return diff result, because on error not executing `$response::transform()` method.

**Comment by dimitrovv** [comment]
@tkvedaras the gateway response should not be transformed if response status is `Error`. If you receive a response with status `error`, you need to handle the thrown exception and to use the error message. 
I guess you have noticed that an exception is being raised when [parsing the gateway response](https://github.com/GenesisGateway/genesis_php/blob/1.5.0/src/Genesis/API/Response.php#L107).
The `error` status usually indicates an invalid request parameters (missed a required parameter) or an improper terminal configuration @ our gateway.

If the status is `declined` then your response should be transformed correctly.

Every transaction type has **required fields** and **conditionally required** (these are being configured per terminal @ our gateway).

If you are processing a `Sale` transaction you have to supply these [required fields](https://github.com/GenesisGateway/genesis_php/blob/1.5.0/src/Genesis/API/Request/Base/Financial/Cards/Synchronous/ASale.php#L78), but sometimes the [billing_address1](https://github.com/GenesisGateway/genesis_php/blob/1.5.0/src/Genesis/API/Request/Base/Financial/Common/Risk/AbstractBaseCustomerInfo.php#L89) can be also required, so this could be a reason for receiving a response with status `error`.

You could use the code below to determine more descriptive error message (`technical_message`):
```php
/**
 * prepare your gateway request here
 */
try {
    $this->gateway->execute();

    $response = $this->gateway->response()->getResponseObject();

    /**
     * do sth with the response
     */
} catch (\Exception $exception) {
    if ($exception instanceof \Genesis\Exceptions\ErrorAPI) {
        $response = $this->gateway->response()->getResponseObject();

        /**
         * use $response->technical_message
         *
         * or
         *
         * use $exception->getMessage() to display the $response->message to the end user
         * $exception->getCode() will be the exact error code from the API Documentation
         */
    }

    /**
     * throw another exception or sth else
     */
}
```

**Comment by tkvedaras** [comment]
Yes I got `error` status with message `'card_number' is invalid`. Invalid card number is not missed field. 
My logs shows that I sent amount multiples 100 times while if status `decline` amount was correct. 

I understand your workflow but I think `getResponseObject()` should return the same format of object all time. If amount is transformed than should be transformed all time or not at all.

**Comment by dimitrovv** [comment]
@tkvedaras I can see now, if your credit card is invalid the status should be set to `declined`, but it depends also on your **gateway / terminal configuration**. This means it can return `error` response if you are trying to process **Authorize3D** with a credit card with **unavailable 3D-Secure**. Your terminal should be configured to process a normal **Authorize** Transaction instead.

Could you provide us the **request xml** and **raw response** and which environment you use (**staging** or **production**)?

Outside of that we will discuss the transformation of the response object in case of `error` status. I think the amount can also be transformed when an error is returned from the gateway.

In your case I am quite sure there is something wrong with the terminal configuration. So if you could provide us the `request xml`, because I couldn't locate your request @ our staging environment.

**Comment by dimitrovv** [comment]
I guess you are using a Merchant Account `Mistertango` for processing through our gateway. All the transactions with **Error class** `InvalidCard` are with status `declined` and the others with `error` status are generated because of an error in the `Workflow` or `InputDataMissing`, but these are transactions in February, all others are processed successfully @ our Staging area.

In order to analyze this better, we will need your **request** and **response** `xml`.

**Comment by tkvedaras** [comment]
Got in production, `unique_id = 453c248cc6252877b5a5f8a33430e3b9`. 
Yes invalid CC number is unavailable 3D ;-)

I think it's nothing bad and safe to transform object all time.

**Comment by dimitrovv** [comment]
@tkvedaras thanks, will be included in **Version 1.6.0** (till end of **April**)

**Comment by tkvedaras** [comment]
Fixed. Thanks
